tests/rptest/tests/retention_policy_test.py

```diff
@@ -130,6 +130,65 @@ def test_changing_topic_retention_with_restart(self, cloud_storage_type):
                                             target_bytes=local_retention)
 
 
+class RetentionPolicyToggleTest(RedpandaTest):
+    topics = (TopicSpec(partition_count=1,
+                        replication_factor=3,
+                        cleanup_policy=TopicSpec.CLEANUP_COMPACT), )
+
+    def __init__(self, test_context):
+        extra_rp_conf = dict(
+            log_compaction_interval_ms=5000,
+            log_segment_size=1048576,
+            compacted_log_segment_size=1048576,
+        )
+
+        super().__init__(test_context=test_context,
+                         num_brokers=3,
+                         extra_rp_conf=extra_rp_conf)
+
+    @cluster(num_nodes=3)
+    def test_changing_topic_cleanup_policy(self):
+        """
+        Test changing topic cleanup policy and retention property,
+        then waits for segments to be removed.
+        """
+
+        self.client().alter_topic_configs(
+            self.topic, {
+                TopicSpec.PROPERTY_CLEANUP_POLICY: TopicSpec.CLEANUP_DELETE,
+            })
+
+        produce_until_segments(
+            self.redpanda,
+            topic=self.topic,
+            partition_idx=0,
+            count=10,
+        )
+
+        local_retention = 10000
+        self.client().alter_topic_configs(
+            self.topic, {
+                TopicSpec.PROPERTY_RETENTION_BYTES: local_retention,
+            })
+
+        wait_for_local_storage_truncate(redpanda=self.redpanda,
+                                        topic=self.topic,
+                                        target_bytes=local_retention)
+
+        self.logger.debug(f"Toggling back to cleanup.policy=compact")
+        self.client().alter_topic_configs(
+            self.topic, {
+                TopicSpec.PROPERTY_CLEANUP_POLICY: TopicSpec.CLEANUP_COMPACT,
+            })
+
+        rpk = RpkTool(self.redpanda)
+        start_offset_before = next(rpk.describe_topic(self.topic)).start_offset
+        produce_total_bytes(self.redpanda, self.topic, 10 * 1024 * 1024)
+        start_offset_after = next(rpk.describe_topic(self.topic)).start_offset
+        assert start_offset_before == start_offset_after,\
```

Comment:
nit: may consider waiting some multiple of `log_compaction_interval_ms` to show that even across iterations of housekeeping, we don't truncate.

Reply:
Updated the test with `time.sleep(5 * self.log_compaction_interval_ms / 1000)`